presto-client/src/main/java/com/facebook/presto/client/SpnegoHandler.java

```diff
@@ -0,0 +1,330 @@
+import static javax.security.auth.login.AppConfigurationEntry.LoginModuleControlFlag.REQUIRED;
+import static org.ietf.jgss.GSSContext.INDEFINITE_LIFETIME;
+import static org.ietf.jgss.GSSCredential.DEFAULT_LIFETIME;
+import static org.ietf.jgss.GSSCredential.INITIATE_ONLY;
+import static org.ietf.jgss.GSSName.NT_HOSTBASED_SERVICE;
+import static org.ietf.jgss.GSSName.NT_USER_NAME;
+
+public class SpnegoHandler
+        implements Interceptor, Authenticator
+{
+    private static final String NEGOTIATE = "Negotiate";
+    private static final Duration MIN_CREDENTIAL_LIFETIME = new Duration(60, SECONDS);
+
+    private static final GSSManager GSS_MANAGER = GSSManager.getInstance();
+
+    private static final Oid SPNEGO_OID = createOid("1.3.6.1.5.5.2");
+    private static final Oid KERBEROS_OID = createOid("1.2.840.113554.1.2.2");
+
+    private final String remoteServiceName;
+    private final boolean useCanonicalHostname;
+    private final Optional<String> principal;
+    private final Optional<File> keytab;
+    private final Optional<File> credentialCache;
+
+    @GuardedBy("this")
+    private Session clientSession;
+
+    public SpnegoHandler(
+            String remoteServiceName,
+            boolean useCanonicalHostname,
+            Optional<String> principal,
+            Optional<File> kerberosConfig,
+            Optional<File> keytab,
+            Optional<File> credentialCache)
+    {
+        this.remoteServiceName = requireNonNull(remoteServiceName, "remoteServiceName is null");
+        this.useCanonicalHostname = useCanonicalHostname;
+        this.principal = requireNonNull(principal, "principal is null");
+        this.keytab = requireNonNull(keytab, "keytab is null");
+        this.credentialCache = requireNonNull(credentialCache, "credentialCache is null");
+
+        kerberosConfig.ifPresent(file -> System.setProperty("java.security.krb5.conf", file.getAbsolutePath()));
+    }
+
+    @Override
+    public Response intercept(Chain chain)
+            throws IOException
+    {
+        // eagerly send authentication if possible
+        try {
+            return chain.proceed(authenticate(chain.request()));
+        }
+        catch (ClientException ignored) {
+            return chain.proceed(chain.request());
+        }
+    }
+
+    @Override
+    public Request authenticate(Route route, Response response)
+            throws IOException
+    {
+        // skip if we already tried or were not asked for Kerberos
+        if (response.request().headers(AUTHORIZATION).stream().anyMatch(SpnegoHandler::isNegotiate) ||
+                response.headers(WWW_AUTHENTICATE).stream().noneMatch(SpnegoHandler::isNegotiate)) {
+            return null;
+        }
+
+        return authenticate(response.request());
+    }
+
+    private static boolean isNegotiate(String value)
+    {
+        return Splitter.on(whitespace()).split(value).iterator().next().equalsIgnoreCase(NEGOTIATE);
```

Comment:
You're right that it can be anywhere. I re-read the RFCs and `WWW-Authenticate` can have multiple challenges